### PR TITLE
Add support for event generators

### LINKED ISSUE #2: Send event with mock data
Here's the repo with the JSON schema for 1st party events:
https://github.com/Azure/azure-rest-api-specs/tree/master/specification/eventgrid/data-plane

This is all about being able to debug your event grid subscriptions

### PR DESCRIPTION
**Pull request by EricJizbaMSFT**
See the changes to the README for more information (and I'm happy to demo to you). I didn't add any gifs/screenshots/tests yet - I figure the user experience might need a little time to bake before I invest too heavily in that stuff.

Fixes #2

### CODE REVIEW

File: src/utils/fsUtils.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        throw new Error(localize('failedUnique', 'Failed to find unique name for new file.'));
+    }
+
+    export function isPathEqual(fsPath1: string, fsPath2: string): boolean {
+        return path.relative(fsPath1, fsPath2) === '';
+    }
+
+    export async function showNewFile(data: string, fileNamePrefix: string, fileNameSuffix: string, viewColumn?: number): Promise<void> {
```
Comment (StephenWeatherford) [inline src/utils/fsUtils.ts]: I recently improved this code a bit...

export async function showNewFile(data: string, extensionPath: string, fileName: string, fileExtension: string, column?: vscode.ViewColumn): Promise<void> {
    let uri: vscode.Uri;
    const folderPath: string = vscode.workspace.rootPath || extensionPath;
    const fullFileName: string | undefined = await getUniqueFileName(folderPath, fileName, fileExtension);
    uri = vscode.Uri.file(path.join(folderPath, fullFileName)).with({ scheme: 'untitled' });
    const textDocument = await vscode.workspace.openTextDocument(uri);
    const editor = await vscode.window.showTextDocument(textDocument, column ? column > vscode.ViewColumn.Three ? vscode.ViewColumn.One : column : undefined, true);
    await writeToEditor(editor, data);
}

File: src/eventSubscription/registerEventSubscriptionCommands.ts [hunk 2]
```diff
@@ -17,9 +21,14 @@ export function registerEventSubscriptionCommands(): void {
     ext.context.subscriptions.push(ext.eventSubscriptionTree);
     ext.context.subscriptions.push(vscode.window.registerTreeDataProvider('azureEventGridSubscriptionExplorer', ext.eventSubscriptionTree));
 
+    ext.context.subscriptions.push(vscode.languages.registerCodeLensProvider({ language: 'json', pattern: '**/*.[eE][vV][eE][nN][tT][gG][eE][nN][eE][rR][aA][tT][oO][rR].[jJ][sS][oO][nN]' }, new MockCodeLensProvider()));
```
Comment (StephenWeatherford) [inline src/eventSubscription/registerEventSubscriptionCommands.ts]: Should you support jsonc?

File: src/eventSubscription/commands/mock/generateEvents.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    if (eventGenerator.jsonSchemaFakerOptions) {
+        // tslint:disable-next-line:no-unsafe-any
+        jsf.option(eventGenerator.jsonSchemaFakerOptions);
+    }
+
+    const events: Promise<{}>[] = [];
+    let count: number = 0;
+    while (count < eventGenerator.numberOfEvents) {
+        // tslint:disable-next-line:no-unsafe-any
+        events.push(jsf.resolve(eventGenerator.eventSchema));
```
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply src/eventSubscription/commands/mock/generateEvents.ts]: The generated events are not identical. They have random data that is generated each time (new guids, different times, different words, etc.)

File: src/utils/fsUtils.ts (hunk 1 above)
Comment (EricJizbaMSFT) [reply src/utils/fsUtils.ts]: Uhh I can't tell what's different and I didn't just copy this code straight from Cosmos so I can't just paste what you have. What did you improve?

File: src/eventSubscription/registerEventSubscriptionCommands.ts (hunk 2 above)
Comment (EricJizbaMSFT) [reply src/eventSubscription/registerEventSubscriptionCommands.ts]: Done. I think the glob pattern is super ugly, but it works for me and I couldn't find anything cleaner. Let me know if you know a better way

File: src/eventSubscription/commands/mock/createMockEventGenerator.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            alwaysFakeOptionals: true
+        },
+        eventSchema: eventSchema
+    };
+
+    await fsUtils.showNewFile(JSON.stringify(eventGenerator, undefined, 2), node.treeItem.label, '.eventGenerator.json');
+}
+
+function getEventTypeFromTopic(topic: string): EventType {
+    if (!!/^\/subscriptions\/[^\/]+$/i.exec(topic)) {
```
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply src/eventSubscription/commands/mock/createMockEventGenerator.ts]: Fixed

File: src/utils/fsUtils.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import { Position, TextDocument, TextEditor, TextEditorEdit, Uri, ViewColumn, window, workspace } from "vscode";
+import { localize } from './localize';
+
+export namespace fsUtils {
+    export async function getUniqueFileName(folderPath: string, prefix: string, suffix?: string): Promise<string> {
+        let count: number = 0;
+        const maxCount: number = 1024;
+
+        while (count < maxCount) {
+            const fullFileName: string = `${prefix}${count === 0 ? '' : `-${count.toString()}`}${suffix}`;
```
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply src/utils/fsUtils.ts]: Got rid of 'toString'

For the other part, I think we'll have to agree to disagree on this stylistic choice. I see your point, but I think single-use variables also negatively affect readability, which is why I chose to do what I did

**Comment by PrashanthCorp** [comment]
Looks fine by me. Onto you @StephenWeatherford

File: resources/templates/Resource.json [hunk 6]
```diff
@@ -0,0 +1,57 @@
+{
+    "description": "Schema of the Data property of an EventGridEvent for a Microsoft.Resources event with additional metadata for generating mock values.",
```
Comment (StephenWeatherford) [inline resources/templates/Resource.json]: nit: Usage of "EventGridEvent" vs "Event Grid event" is inconsistent in these jsons

**Pull request merged**